contracts/schemes/Competition.sol

```diff
@@ -0,0 +1,349 @@
+        suggestions[suggestionsCounter].proposalId = _proposalId;
+        suggestions[suggestionsCounter].suggester = msg.sender;
+        emit NewSuggestion(_proposalId, suggestionsCounter, _descriptionHash, msg.sender);
+        return suggestionsCounter;
+    }
+
+    /**
+    * @dev vote on a suggestion
+    * @param _suggestionId suggestionId
+    * @return bool
+    */
+    function vote(uint256 _suggestionId)
+    external
+    returns(bool)
+    {
+        bytes32 proposalId = suggestions[_suggestionId].proposalId;
+        require(proposalId != bytes32(0), "suggestion not exist");
+        setSnapshotBlock(proposalId);
+        Proposal storage proposal = proposals[proposalId];
+        // solhint-disable-next-line not-rely-on-time
+        require(proposal.endTime > now, "competition ended");
+        Suggestion storage suggestion = suggestions[_suggestionId];
+        require(suggestion.votes[msg.sender] == 0, "already voted on this suggestion");
+        require(proposal.votesPerVoter[msg.sender] < proposal.numberOfVotesPerVoters,
+        "exceed number of votes allowed");
+        proposal.votesPerVoter[msg.sender] = proposal.votesPerVoter[msg.sender].add(1);
+        suggestion.votes[msg.sender] = suggestion.votes[msg.sender].add(1);
+        Avatar avatar = ContributionRewardExt(proposal.contributionRewardExt).avatar();
+        uint256 reputation = avatar.nativeReputation().balanceOfAt(msg.sender, proposals[proposalId].snapshotBlock);
+        require(reputation > 0, "voter has no reputation");
+        suggestion.totalVotes = suggestion.totalVotes.add(reputation);
+        suggestion.votes[msg.sender] = reputation;
+        proposal.tiesSuggestions[suggestion.totalVotes] = proposal.tiesSuggestions[suggestion.totalVotes].add(1);
+        refreshTopSuggestions(proposalId, _suggestionId);
+        emit NewVote(proposalId, _suggestionId, msg.sender, reputation);
+        return true;
+    }
+
+    /**
+    * @dev redeem a winning suggestion reward
+    * @param _suggestionId suggestionId
+    * @param _beneficiary - the reward beneficiary.
+    *        this parameter is take into account only if the msg.sender is the suggestion's suggester,
+    *        otherwise the _beneficiary param is ignored and the beneficiary is suggestion's suggester.
+    */
+    function redeem(uint256 _suggestionId, address payable _beneficiary) external {
+        address payable beneficiary = suggestions[_suggestionId].suggester;
+        if (msg.sender == suggestions[_suggestionId].suggester) {
+            //only suggester can redeem to other address
+            if (beneficiary != address(0)) {
+                beneficiary = _beneficiary;
+            }
+        }
+        _redeem(_suggestionId, beneficiary);
+    }
+
+    /**
+    * @dev setSnapshotBlock set the block for the reputaion snapshot
+    * @param _proposalId the proposal id
+    */
+    function setSnapshotBlock(bytes32 _proposalId) public {
```

Comment:
yes. 
-it can be called by any one 
-it is called on the first vote
-the block number can be set. one time
-it cannot be set before the voting time due to https://github.com/daostack/arc/blob/new_cr/contracts/schemes/Competition.sol#L249